### PR TITLE
Teleportation demo

### PR DESCRIPTION
**Pull request by zacespinosa**
Can you look it through and make sure nothing is wrong? The threading stuff is a bit confusing, but I think I checked for all locking situations.

### CODE REVIEW

File: qnetdes/distributedGates.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+def notify_entangler_is_done(caller, target_agents): 
+    for agent in target_agents:
+        cbit = [1] # cbit signaling done
+        caller.csend(agent.name, cbit)
+
+def cat_entangler(control, targets, entangled=False, notify=False):
+    '''
+    Performs the cat entangler, one of two primitive operations for 
+    distributed quantum computing which can be used to implement non-local operations.
+    Projects the state Alice's local control bit (phi) on entangled qubits owned by other Agents,
```
Comment (mattradz) [inline qnetdes/distributedGates.py]: "the state of"

File: qnetdes/distributedGates.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        caller.csend(agent.name, cbit)
+
+def cat_entangler(control, targets, entangled=False, notify=False):
+    '''
+    Performs the cat entangler, one of two primitive operations for 
+    distributed quantum computing which can be used to implement non-local operations.
+    Projects the state Alice's local control bit (phi) on entangled qubits owned by other Agents, 
+    allowing Agents to effectively use Alice's qubit as a control for their own operations. Measurements
+    are stored in ro in the position corresponding to qubit's index (i.e. ith qubit measured into ro[i]). 
+    Once the cat_entangler is started in the control agent, each target agent to wait until they have received
```
Comment (mattradz) [inline qnetdes/distributedGates.py]: "waits"

File: qnetdes/distributedGates.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+from pyquil.api import WavefunctionSimulator, QVMConnection
+from pyquil.gates import *
+from qnetdes import *
+
+__all__ = ["cat_entangler", "cat_disentangler"]
+
+def distributed_gate(agent):
+    agent.using_distributed_gate = not agent.using_distributed_gate
+
+def notify_entangler_is_done(caller, target_agents):
```
Comment (mattradz) [inline qnetdes/distributedGates.py]: Do they need a crecv if notify is on?

Comment (zacespinosa) [reply qnetdes/distributedGates.py]: yeah, they should be expecting a cbit

**Pull request merged**